docs/udev-configuration.md

```diff
@@ -8,13 +8,9 @@ Akri's udev discovery handler parses udev rules listed in a Configuration, searc
 list of discovered device nodes (ie: /dev/video0). You tell Akri which device(s) to find by passing [udev
 rules](https://wiki.archlinux.org/index.php/Udev) into a Configuration. Akri has created a
 [grammar](../agent/src/protocols/udev/udev_rule_grammar.pest) for parsing the rules, expecting them to be formatted
-according to the [Linux Man pages](https://linux.die.net/man/7/udev). While udev rules are normally used to both find
+according to the [Linux Man pages](https://man7.org/linux/man-pages/man7/udev.7.html). While udev rules are normally used to both find
 devices and perform actions on devices, the Akri udev discovery handler is only interested in finding devices.
-Consequently, the discovery handler will ignore all action operations ("=" , "+=" , "-=" , ":=") and action fields such
-as `TEST` in the udev rules. You should only use match operations ("==",  "!=") and fields such as `KERNEL` and
-`DEVPATH` in your udev rules. Additionally, there are some match fields that look up the device hierarchy, such as
-`SUBSYSTEMS`, that are yet to be supported and are also ignored. See the
-[grammar](../agent/src/protocols/udev/udev_rule_grammar.pest) for the full list of supported fields.
+Consequently, the discovery handler will throw an error if any of the rules contain an action operation ("=" , "+=" , "-=" , ":=") or action fields such as `IMPORT` in the udev rules. You should only use match operations ("==",  "!=") and fields such as `KERNEL` and `DEVPATH` in your udev rules. Additionally, there are some match fields that look up the device hierarchy, such as `SUBSYSTEMS`, that are yet to be supported and will throw an error if used. Support for these will be added soon. To summarize, only the following udev fields are currently supported: `DEVPATH`, `KERNEL`, `TAG`, `DRIVER`, `SUBSYSTEM`, `ATTRIBUTE`, `PROPERTY`.
```

Comment:
Consider pointing to the example*.rules files from here as well.

Reply:
Also, can we add a line that would point to some docs on how to validate the rules? My worry is the following:
1) User identifies a device that is to be matched
2) User uses udevadm to list its fields and constructs the UDEV filter
3) User does not know if the filter is too wide, such that it would include other unintended devices
4) User deploys Akri Config and is surprised that it hit more than what was expected

udevadm is great for figuring out the attributes specific to a device, but does not help to ensure that you are not casting a wide net (unless there is some other mode that I missed).

Reply:
This is a good point. While `udevadm trigger` supports looking up devices based on the currently supported fields, it does not support looking up the device hierarchy, so you cannot use it to test `ATTRS` when support comes for that, and it does not support regex like udev. Adding in more details as to how to use `udevadm trigger` would only be a temporary and incomplete solution to helping people craft udev rules. As an example, to test the udev rule `--set udev.udevRules[0]='KERNEL=="video1"\, ATTRS{manufacturer}=="Some Manufacturer"'`, you could run
`udevadm trigger --verbose --dry-run --type=devices --name-match=video1 --attr-match=manufacturer="Some Manufacturer"`

A different solution is the following:
Create a new rules file in the `/etc/udev/rules.d/` directory. There should already be the `70-snap.core.rules` file there that is fun to explore. Add the rule[s] you are planning to add to your configuration to it, adding an TAG action to the rule. Then you can search for all devices with that tag. 
1. navigate to `/etc/udev/rules.d`
1. Create a file called `90-akri.rules` and add your rules to it. Note how 90 is the prefix to the file name. This makes sure these rules are run before the others in `70-snap.core.rules`, preventing them from being overwritten.
```sh
# Rule to tag video and sound devices
KERNEL=="video[0-9]*", TAG+="akri_tag"
SUBSYSTEM=="sound", KERNEL=="card[0-9]*", TAG+="akri_tag"
```
1. Reload the udev rules and trigger them
```sh
udevadm control --reload
udevadm trigger
```
1. List the devices that have been tagged, which Akri will discover
```sh
udevadm trigger --verbose --dry-run --type=devices --tag-match=akri_tag
```
1. Modify rule as needed, being sure to reload and trigger the rules each time
1. Remove the tag from devices
```sh
# Rule to tag video and sound devices. Note how  `+=` turns to `-=`.
KERNEL=="video[0-9]*", TAG-="akri_tag"
SUBSYSTEM=="sound", KERNEL=="card[0-9]*", TAG-="akri_tag"
```
1. Reload the udev rules and trigger them
```sh
udevadm control --reload
udevadm trigger
```
1. Confirm the tag has been removed and no devices are listed
```sh
udevadm trigger --verbose --dry-run --type=devices --tag-match=akri_tag
```
1. Create an Akri configuration with your udev rule!

Reply:
I presume the first example that you mentioned should not have been `ATTRS{manufacturer}`, but instead `ATTR{manufacturer}`, right? 

I think this is a great guide, we should include it in the docs.

Also, could the tagging approach be used to at least workaround the missing support for parent features? It would mean that somebody has to deploy the udev rule file and refresh udev rules, but then would we be able to just use Akri to detect the specific tag? 

In my case, this is all I get for the leaf device, which is not super useful. Only parent devices have useful attributes.
```
looking at device '/devices/platform/soc/3f980000.usb/usb1/1-1/1-1.5/1-1.5:1.0/ttyUSB0/tty/ttyUSB0':
    KERNEL=="ttyUSB0"
    SUBSYSTEM=="tty"
    DRIVER==""
```

Reply:
Double yes: accidentally added the S again and yes tagging could be a work around

Reply:
Added documentation to `udev-configuration.md`. Let me know what you think.

Reply:
i might simplify this a bit ... from:

You should only use match operations ("==",  "!=") and fields such as `KERNEL` and `DEVPATH` in your udev rules. Additionally, there are some match fields that look up the device hierarchy, such as `SUBSYSTEMS`, that are yet to be supported and will throw an error if used. Support for these will be added soon. To summarize, only the following udev fields are currently supported: `DEVPATH`, `KERNEL`, `TAG`, `DRIVER`, `SUBSYSTEM`, `ATTRIBUTE`, `PROPERTY`.

to:

You should only use match operations ("==",  "!=") and the following udev fields: `DEVPATH`, `KERNEL`, `TAG`, `DRIVER`, `SUBSYSTEM`, `ATTRIBUTE`, `PROPERTY`.  There are some match fields that look up the device hierarchy, such as `SUBSYSTEMS`, that are yet to be supported and will throw an error if used. Support for these will be added soon.